### PR TITLE
_eval_rewrite_as_gamma for factorial2

### PR DESCRIPTION
**Pull request by meghana-madhyastha**
I have extended the factorial 2 function to work for complex numbers as well (The current behavior supports only real numbers). I have added a test as well. It would be good if someone can give me feedback on this.

### CODE REVIEW

**Comment by jksuom** [comment]
A function defined for integers has an infinite number of extensions to the complex domain. Is there a reason for choosing this one?

**Comment by meghana1995** [comment]
Currently, sympy only supports double factorials for real numbers. I thought it would make sense to extend it to the complex domain. (Functions such as factorial can be generalised to the complex domain using the gamma functions)
It was originally suggested by J. Keiper and O. I. Marichev (1994) during the implementation of the function Factorial2 in Mathematica. 

Here are some of the references I followed for the same:
mathworld.wolfram.com/DoubleFactorial.html
http://functions.wolfram.com/GammaBetaErf/Factorial2/introductions/FactorialBinomials/ShowAll.html
http://www.maplesoft.com/support/help/maple/view.aspx?path=factorial

**Comment by meghana1995** [comment]
The tests for pretty printing seem to be failing because initially, factorial2( factorial2(factorial2(n)) ) didn't evaluate for an integer symbol n. So the output is not ((n!!)!!)!! as it used to be. It instead computes
2 ** ((1 + 2 * arg - cos(pi * arg)) / 4) * (pi ** ((cos(pi * arg) - 1) / 4)) * gamma (1 + arg / 2)
which is the expression for the double factorial of a complex number.
Should I remove these pretty printing tests or should I modify them so that they check with the new result ?
@aktech @jksuom

**Comment by jksuom** [comment]
I would prefer not to change the printing test. Instead, I think that if factorial2 is extended to the complex domain, it should also have a method for evaluating is_real.

**Comment by meghana1995** [comment]
I thought of adding an _eval_rewrite_as_gamma function to factorial2 (the same way it has been done for factorial and other classes). This way, if one wants to evaluate factorial2 for a complex number, all one has to do is call the rewrite function and substitute the complex number which one wants to evaluate. For example something along the lines of:
n = Symbol('n', complex=True)
func = factorial2(n).rewrite(gamma) 
func.subs(n, 1 + 2*I).evalf(10)
I have added this as a unit test. I thought that this is probably a better idea than extending it to the complex domain as it has not been done with the other factorial functions. Let me know if I am wrong.

**Comment by meghana1995** [comment]
@jksuom, should I revert back to the old commit where I directly extended factorial2 to the complex domain, or does the function _eval_rewrite_as_gamma () make more sense as it is consistent with what has been done for the other combinatorial functions (such as factorial(), risingfactorial(), etc) ? What is your opinion ?

**Comment by jksuom** [comment]
I think the rewrite version is better. It can be used to compute factorial2 for integer arguments. Only, I believe this could achieved  with other, simpler. expressions. The values of factorial2 depend on the parity of the argument as follows:
```
n!! = 2**(n/2)*gamma(n/2 + 1)             # for even n
n!! = sqrt(2/pi)*2**(n/2)*gamma(n/2 + 1)  # for odd n
```
These values can be combined in the formula
``n!! = f(n)*2**(n/2)*gamma(n/2 + 1)``
where ``f`` is any function such that ``f(n) = 1`` for even ``n`` and ``f(n) = sqrt(2/pi)`` for odd ``n``.

The simplest such functions are probably periodic. The function suggested by Keiper and Marichev is
``f(n) = (2/pi)**((1 - cos(pi*n))/4)``
which is periodic of period 2, but there other periodic functions with the same values.

The Fourier series of the above function has an infinite number of terms since the cosine is in the exponent. A periodic function with the least number of Fourier components would be
``f(n) = (1 + cos(pi*n))/2 + sqrt(2/pi)*(1 - cos(pi*n))/2``
(unless I made a mistake in the computation).
This might be computationally more effective than a function with a cosine in the exponent. Though I have not made any tests.

**Comment by meghana1995** [comment]
@jksuom Using f(n) = (1 + cos(pi*n))/2 + sqrt(2/pi)*(1 - cos(pi*n))/2 would definitely be computationally more effective than f(n) = (2/pi)**((1 - cos(pi*n))/4).
However, when the formula f(n) = (2/pi)**((1 - cos(pi*n))/4) is used, the rewrite function can be used to compute factorial2 for integer as well as complex arguments. When I use f(n) = (1 + cos(pi*n))/2 + sqrt(2/pi)*(1 - cos(pi*n))/2, the tests for _rewrite_as_gamma for complex args are failing. Should I change the function f(n) as well as the  tests to include only integer arguments and restrict the functionality to integer arguments ?

**Comment by jksuom** [comment]
We must remember that factorial2  is defined for integer arguments only. So its values for general complex arguments are undefined.

It is true that there are complex (meromorphic) functions that agree with factorial2 on the set of integers. But there is a great number of such functions, and there appears to be no single preferred one. The formula presented by Keiper and Marichev has a short expression, but nothing indicates that it would be useful for other than integer arguments. Hence it does not make sense to speak of the values of factorial2 for complex arguments.

As to a possible rewrite formula I am inclined to think that a ``Piecewise`` expression would be the most appropriate one. It would have one expression (``1``) for even arguments and another (``sqrt(2/pi)``) for odd arguments. And there would probably be no need for additional argument tests.

**Comment by meghana1995** [comment]
@jksuom , I have modified _eval_rewrite_as_gamma so that it uses a Piecewise expression as you had suggested. Can you review this ?

**Comment by jksuom** [comment]
This looks good. Just a couple of remarks:

I suspect that the condition ``Eq(Mod(n, 2), 0)`` would often remain unevaluated if ``n`` is an expression that is not quite simple. Probably ``n.is_even`` would be better. (And similarly with ``n.is_odd``.)

It is recommended that imports are put at the top of the file.

The spacing in the rewrite formula could be the same as that produced by the printer, e.g, ``n/2`` instead of ``n / 2``, etc.

Finally, please change the name of this PR to show its actual content.

**Comment by meghana1995** [comment]
@jksuom ,
I moved some of the imports to the top. However, with some functions, there were import errors when I tried moving them to the top of the file.

The problem with n.is_even is that, it either returns 'None' , 'True' or 'False' so in case it is not known initially whether or not the symbol is even or odd, the Piecewise will evaluate to false even though it would be better if it were to remain in its symbolic form.

For example, if I change it to `return 2**(n/2)*gamma(n/2 + 1) * Piecewise((1, n.is_even==True), (sqrt(2/pi), n.is_odd==True))` 

`>> n = Symbol('n', integer=True)`
`>> factorial2(n).rewrite(gamma)`

`2**(n/2)*Piecewise()*gamma(n/2 + 1)`

**Comment by asmeurer** [comment]
Yeah, the use of Mod is better. is_even is only for direct evaluation.

**Comment by jksuom** [comment]
I believe this rewrite would be principally useful for direct evaluation. But it appears ``Mod`` should be used because of those problems.

File: sympy/functions/combinatorial/factorials.py [hunk 1]
```diff
@@ -324,7 +327,9 @@ class factorial2(CombinatorialFunction):
 
     @classmethod
     def eval(cls, arg):
+        from sympy import sin, cos, gamma, pi
```
Comment (jksuom) [inline sympy/functions/combinatorial/factorials.py]: This seems to be a new line. Is it necessary?

Comment (jksuom) [inline sympy/functions/combinatorial/factorials.py]: Is it here for import errors?

**Comment by meghana1995** [comment]
@jksuom , I think I had added this line before and forgot to delete it. Sorry for that. I removed it now.

**Comment by meghana1995** [comment]
@jksuom The travis build seems to be failing and the error seems to be
  No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
Is it possible to restart the build ?

**Comment by jksuom** [comment]
Yes, I've restarted the jobs.

**Comment by jksuom** [comment]
Thanks! I think this is ready. Merging.

**Pull request merged**